### PR TITLE
FUSETOOLS2-444 - reorganization of the ui for creating an environment…

### PR DESCRIPTION
**Pull request by lhein**
… variable from ConfigMap/Secret

Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

### CODE REVIEW

**Comment by apupier** [comment]
is the intent to refactor/clean the code that will be mainly touched before implementing the feature?

**Comment by lhein** [comment]
> I don't understand the purpose of this PR.

Taken from the Jira issue:
This task is only handling the reorganization of the UI part of the creation process of an environment variable reference. When creating a reference we want to ask only for the reference name, the variable name and the value. If the reference file exists on file system it will be used to provide a list of predefined env variables for the users to select from. There should be also an entry in that list which allows the user to create a new variable instead of selecting one.

What exactly is not clear to you?

**Comment by apupier** [comment]
I don't see the difference in the UI between current and this PR. Can you point it?

**Comment by lhein** [comment]
> I don't see the difference in the UI between current and this PR. Can you point it?

This only affects creating a variable from a Secret / ConfigMap. We ask now for the reference name first instead of the variable name. Also we are no longer asking for a reference key because now we reuse the variable name. I also made preparations for a quickpick for selecting existing variables from an existing file or to select a 'New' entry in the list if I still want to create a new variable instead of selecting an existing. The implementation of the logic for checking for the local secret/configmap file and to process it in terms of reading the contained variables will be part of the second subtask.

File: src/commands/CreateEnvironmentVariableCommand.ts [hunk 1]
```diff
@@ -21,25 +21,27 @@ import * as vscode from 'vscode';
 import { DataVirtConfig, ConfigMapRef, SecretRef, Property, KeyRef } from '../model/DataVirtModel';
 import { EnvironmentTreeNode } from '../model/tree/EnvironmentNode';
 
+const CREATE_NEW_ENTRY: string = '...New...';
```
Comment (apupier) [inline src/commands/CreateEnvironmentVariableCommand.ts:24]: Seems strange to have the first 3 dots. Why not just 'New..'?

File: src/commands/CreateEnvironmentVariableCommand.ts [hunk 2]
```diff
@@ -54,46 +56,53 @@ export async function createEnvironmentVariableCommandForConfigMap(envNode: Envi
...
 		}
 	}
 }
 
 
-export function handleEnvironmentVariableCreation(dvConfig: DataVirtConfig, enviroment: Property[], file: string, entryType: string, entryName: string, entryValue: string, refName?: string, refKey?: string): Promise<boolean> {
+export function handleEnvironmentVariableCreation(dvConfig: DataVirtConfig, enviroment: Property[], file: string, variableType: string, variableName: string, variableValue: string, refName?: string): Promise<boolean> {
 	return new Promise<boolean>( (resolve) => {
-		if (dvConfig && entryType && enviroment && file && entryName) {
+		if (dvConfig && variableType && enviroment && file && variableName) {
```
Comment (apupier) [inline src/commands/CreateEnvironmentVariableCommand.ts:100]: optional as not part of this PR: typo enviroment

File: src/commands/CreateEnvironmentVariableCommand.ts [hunk 3]
```diff
@@ -104,24 +113,50 @@ export function handleEnvironmentVariableCreation(dvConfig: DataVirtConfig, envi
...
+
+async function queryVariable(envNode: EnvironmentTreeNode, predefinedVariables: Array<Property>): Promise<Property | undefined> {
+	const names: string[] = new Array<string>();
+	names.push(CREATE_NEW_ENTRY);
+
+	predefinedVariables.forEach( (variable: Property) => {
+		names.push(variable.name);
+	});
+
+	const selectedVariableName: string = await vscode.window.showQuickPick(names, { canPickMany: false, placeHolder: `Select a variable from the list or enter a new one` });
```
Comment (apupier) [inline src/commands/CreateEnvironmentVariableCommand.ts:144]: the text says to choose one or to enter a new one but it is showQuickpick so I cannot type a new one. When typing it is only filtering available items in the list.

Comment (apupier) [inline src/commands/CreateEnvironmentVariableCommand.ts:144]: shouldn't we have a validation on the name that can be provided?

File: src/utils.ts [hunk 4]
```diff
@@ -157,7 +157,19 @@ export function openDDLEditor(vdbName: string) {
...
 
-export function createOrUpdateLocalReferenceFile(refName: string, refKey: string, entryValue: string, entryType: string) {
+export function createOrUpdateLocalReferenceFile(refName: string, variableName: string, variableValue: string, refType: string) {
 	// TODO: create or update the entry reference in a local yaml file for configMap OR secret format
 	// TODO: implement me!
 }
+
+export function doesLocalReferenceFileExist(vdbFile: string, refName: string, refType: string): boolean {
+	// TODO: check if a file exists for the given type and name in the same folder than the vdbFile
+	// TODO: implement me!
```
Comment (apupier) [inline src/utils.ts:167]: the line `TODO implement me` is not very useful. The TODO implies that it would be better to be implemented already.

**Comment by lhein** [comment]
>     * not a big fan of merging non-functional UI in master. It means that master is not releasable but fine for this one as it was already the case for this part of code, it is not worse than before from a "releasable state" point of view.

It provides the same functionality as already is in master. Not sure what you are referring to. We don't want to release it before all of the required jiras have been solved. There is simply no need for that.

File: src/commands/CreateEnvironmentVariableCommand.ts (hunk 1 above)
Comment (lhein) [reply src/commands/CreateEnvironmentVariableCommand.ts:24]: That was more a protective measure as I wanted to ensure that this item is always on top. I haven't tested yet if the quickpicks are sorted in some way. If that is not the case then I will rename it to just "New...".

File: src/commands/CreateEnvironmentVariableCommand.ts (hunk 3 above)
Comment (lhein) [reply src/commands/CreateEnvironmentVariableCommand.ts:144]: Agreed. That needs to be `Select New...` instead.

Comment (lhein) [reply src/commands/CreateEnvironmentVariableCommand.ts:144]: It is a quick pick. If those values are part of an existing ConfigMap or Secret I would consider them as valid. 
@rareddy Any opinion on that matter? Do we need an additional check for invalid Secret / ConfigMap contents not created by our DV extension?

File: src/utils.ts (hunk 4 above)
Comment (lhein) [reply src/utils.ts:167]: Its a comment Aurelien. Only developers will see them and they will disappear once it got implemented.

Comment (apupier) [reply src/utils.ts:167]: it remains better to not write in the first place

File: src/commands/CreateEnvironmentVariableCommand.ts (hunk 1 above)
Comment (apupier) [reply src/commands/CreateEnvironmentVariableCommand.ts:24]: ok, understand the use case.

I'm surprised that VS Code doesn't allow that. I've seen several places with group separators that would do the trick but... it is not part of API!!! https://github.com/microsoft/vscode/issues/74967
a specific sort seems not possible either: https://github.com/microsoft/vscode/issues/63050

**Comment by apupier** [comment]
> > ```
> > * not a big fan of merging non-functional UI in master. It means that master is not releasable but fine for this one as it was already the case for this part of code, it is not worse than before from a "releasable state" point of view.
> > ```
> 
> 
> It provides the same functionality as already is in master. Not sure what you are referring to. We don't want to release it before all of the required jiras have been solved. There is simply no need for that.

I'm referring to the split of the task in general. Providing non-functional code in the master is setting the master in a non-releasable state. Given that VS Code doesn't allow to do maintenance release easily it is even more important to be in a releasable state.

**Comment by lhein** [comment]
I have split that work into two separate issues to have smaller and easier to review PRs which was one of your major complains so far whenever I added a new feature to this project in a bigger PR. 

I'd suggest you think about what you really want...smaller and easy to review PRs OR bigger PRs which take longer to review but contain the full feature.

File: src/commands/CreateEnvironmentVariableCommand.ts (hunk 1 above)
Comment (lhein) [reply src/commands/CreateEnvironmentVariableCommand.ts]: I renamed it to "New..." and will revisit it when doing the logic part for filling that quickpick.

**Comment by apupier** [comment]
> I have split that work into two separate issues to have smaller and easier to review PRs which was one of your major complains so far whenever I added a new feature to this project in a bigger PR.

this is great

> I'd suggest you think about what you really want...smaller and easy to review PRs OR bigger PRs which take longer to review but contain the full feature.

I still want smaller and easy PR to review which doesn't introduce non-working bits available to users. The "full feature" can be split in a lot of smaller ones. For instance in this case, other possible splits are:
- split 1:
  - write first the backend methods with tests for backend methods for the "full feature"
  - write UI reusing the backend methods with test for UI flow for the "full feature"
- split 2:
  - write UI first with a toggle to not show it to users with test for UI flow mocking backends for the "full feature"
  - write backend methods removing mock for backend of test and the toggle for the "full feature"
- split 3:
  - write backend and UI with tests to choose values from an existing properties (configmap or secret)
  - write backend and UI with tests to allow creating a new key in an existing properties file
  - write backend and UI with tests to allow providing a value to a created key in an existing properties file
  - write backend and UI with tests to allow providing a key and value for an unexisting properties file

and there are surely a lot of other possibilities which allows to always have a releasable master while trying to keep the PR as small as possible

**Comment by lhein** [comment]
> I still want smaller and easy PR to review which doesn't introduce non-working bits available to users

Why do you keep repeating this false statement over and over? It doesn't become true by doing so.

The master branch is functional and this PR doesn't change that. I also mentioned before that releasing at this point is nothing we gonna even consider.

**Comment by apupier** [comment]
> > I still want smaller and easy PR to review which doesn't introduce non-working bits available to users
> 
> Why do you keep repeating this false statement over and over? It doesn't become true by doing so.
> 
> The master branch is functional and this PR doesn't change that.

the master branch is not fully functional, there are functionalities displayed in the UI that are not working.

>  I also mentioned before that releasing at this point is nothing we gonna even consider.

I said that it is fine for this PR as it was already not functional before anyway. So we both agree.
I wanted to raise the points to try to improve future work.

**Comment by lhein** [comment]
> the master branch is not fully functional, there are functionalities displayed in the UI that are not working.
> 
> >

I'd say raise a Jira if you discovered a bug but this is really out of scope of this PR.

**Comment by apupier** [comment]
> > the master branch is not fully functional, there are functionalities displayed in the UI that are not working.
> > >
> 
> I'd say raise a Jira if you discovered a bug but this is really out of scope of this PR.

this was already raised https://issues.redhat.com/browse/FUSETOOLS2-332?focusedCommentId=14065939&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14065939

**Comment by lhein** [comment]
and I gave you an explanation on that as well:
image

Calling the master "not functional" is wrong. If you don't want to approve the PR that's fine for me, but stop wasting my time by causing an endless discussion loop.

**Comment by apupier** [comment]
> Calling the master "not functional" is wrong.

the data provided by the user is lost. This is a big bug. The feature provided in the UI is not functional.

> If you don't want to approve the PR that's fine for me, but stop wasting my time by causing an endless discussion loop.

I mentioned this point as not ideal and said that it is fine. So this is not a blocking point on this PR.

File: src/commands/CreateEnvironmentVariableCommand.ts (hunk 3 above)
Comment (lhein) [reply src/commands/CreateEnvironmentVariableCommand.ts]: I am taking this discussion to the next dv tooling call. If required this will be addressed in a separate jira.

**Comment by lhein** [comment]
closing this one because of a weird test issue on travis. created #166
